### ISSUE TITLE
Bundler reports missing native extensions which are there

### ISSUE THREAD

**Issue by jeremywadsack** [bundler/bundler#6508]
# Error Report

## Questions

Please fill out answers to these questions, it'll help us figure out why things are going wrong.

- **What did you do?**
  I ran `bin/rails console` and bundler reported a missing shared object file. e.g. 
  ```
  1: from /home/app/query-processing/vendor/bundle/ruby/2.5.0/gems/oj-3.5.0/lib/oj.rb:21:in `<top (required)>'
  /home/app/query-processing/vendor/bundle/ruby/2.5.0/gems/oj-3.5.0/lib/oj.rb:21:in `require': libruby.so.2.5: cannot open shared object file: No such file or directory - /home/app/query-processing/vendor/bundle/ruby/2.5.0/gems/oj-3.5.0/lib/oj/oj.so (LoadError)
  ```

  The file exists and is world readable and executable:
  ```
  $ ls -l /home/app/query-processing/vendor/bundle/ruby/2.5.0/gems/oj-3.5.0/lib/oj/oj.so
  -rwxr-xr-x 1 app app 810824 Apr 18 18:31 /home/app/query-processing/vendor/bundle/ruby/2.5.0/gems/oj-3.5.0/lib/oj/oj.so
  ```

  This happens for one of our Docker images for this app, but not on the other one. In this image we build ruby 2.5.0 using `ruby-install`. In the other image we loaded ruby with rvm. In other projects, we have done this without problems (using ruby 2.4).

  Following a similar report in #3664, I tried updating the gem (`/usr/local/bin/bundle update oj`), which resulted in this report.

- **What did you expect to happen?**

  I expected Bundler to be able to load shared objects that exist.

- **What happened instead?**

  Bundler reported "cannot open shared object file".

- **Have you tried any solutions posted on similar issues in our issue tracker, stack overflow, or google?**

  The first gem that caused this was `nokogiri`. I tried removing this with `gem` and re-installing it.
  ```
  bundle exec gem uninstall nokogiri
  bundle install --without development test --deployment
  ```
  That resolved the issue for `nokogiri` and then raised the error on `oj` (the next gem with native extensions).

  I need to have the gems work the _first_ time I build them (when building the image). And not having to remove and re-install them.

  I tried rebuilding the Docker image from scratch (even removing all cached images) but that doesn't resolve this.

- **Have you read our issues document, https://github.com/bundler/bundler/blob/master/doc/contributing/ISSUES.md?**

  Yes. 
  - I removed user-specific gems and git repos and system-wide git repos.
  - I removed git checkouts with no change in behavior. 
  - I removed project-specific settings and had to re-run `bundle install --without development test --deployment` to get bundler to even recognize the installed files, but that didn't resolve the loading issue.

## Backtrace

```
RuntimeError: Unable to find a spec satisfying resque (~> 1.27) in the set. Perhaps the lockfile is corrupted?
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/spec_set.rb:40:in `block in for'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/spec_set.rb:25:in `loop'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/spec_set.rb:25:in `for'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/spec_set.rb:80:in `materialize'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/definition.rb:170:in `specs'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/definition.rb:158:in `resolve_remotely!'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/installer.rb:279:in `resolve_if_needed'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/installer.rb:83:in `block in run'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/process_lock.rb:12:in `block in lock'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/process_lock.rb:9:in `open'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/process_lock.rb:9:in `lock'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/installer.rb:72:in `run'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/installer.rb:25:in `install'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/cli/update.rb:61:in `run'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/cli.rb:266:in `update'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/cli.rb:27:in `dispatch'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/cli.rb:18:in `start'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/exe/bundle:30:in `block in <top (required)>'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
  /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/exe/bundle:22:in `<top (required)>'
  /usr/local/bin/bundle:23:in `load'
  /usr/local/bin/bundle:23:in `<main>'
```

## Environment

```
Bundler       1.16.1
  Platforms   ruby, x86_64-linux
Ruby          2.5.0p0 (2017-12-25 revision 61468) [x86_64-linux]
  Full Path   /usr/local/bin/ruby
  Config Dir  /usr/local/etc
RubyGems      2.7.3
  Gem Home    /home/app/query-processing/vendor/bundle/ruby/2.5.0
  Gem Path    
  User Path   /home/app/.gem/ruby/2.5.0
  Bin Dir     /home/app/query-processing/vendor/bundle/ruby/2.5.0/bin
OpenSSL       
  Compiled    OpenSSL 1.0.2g  1 Mar 2016
  Loaded      OpenSSL 1.0.2g  1 Mar 2016
  Cert File   /usr/lib/ssl/cert.pem
  Cert Dir    /usr/lib/ssl/certs
Tools         
  Git         2.7.4
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2017-12-21
Git SHA           0034ef341
Released Version  true
```

## Bundler settings

```
frozen
  Set for your local app (/home/app/query-processing/.bundle/config): true
path
  Set for your local app (/home/app/query-processing/.bundle/config): "vendor/bundle"
without
  Set for your local app (/home/app/query-processing/.bundle/config): [:development, :test]
```

## Gemfile

I can provide detail on these through a private channel if needed. I think they aren't related to the core issue (the failure to load shared objects).

**Comment by segiddins** [comment]
I'm guessing the failure to load the shared object file is _because_ it was built with a different version of ruby than you are currently running. Does `bundle pristine oj` fix the issue, by any chance?

**Comment by jeremywadsack** [comment]
@segiddins Thanks for the quick response. Yes, `bundle pristine oj` resolved this. (I think that's effectively what I did with `bundle exec gem uninstall ...; bundle install ...` but faster.)

The ruby version is a good guess, but I don't think that's the issue. I mounted the docker image layer just _before_ we install ruby and there is no ruby on the path.

On further testing, it turns out I was not using a newly built docker image. I was testing in a cached version (in our GKE cluster) which was out-of-date. When I tested with the latest docker image (by image tag), everything worked fine.

I'm guessing this is related to either a corrupt image or something with our CI that is building on a different kernel than the production environment. In any case it's not a bundler issue, so I'm closing this.
